### PR TITLE
Update release process

### PR DESCRIPTION
**Pull request by wipfli**
This pull request reforms the release process.

* Workflow dispatches on manual trigger.
* Tagging by hand is not required anymore. 
* The version is taken from `package.json`.
* All testing is done by `ci.yml`.
* CI on pull requests to the main branch will ensure that we don't publish rubbish to NPM.

### CODE REVIEW

**Comment by HarelM** [comment]
We are not running build-csp in any place?
Are we still using `prepublishOnly` build step somewhere? Seems like there are things that I thought are running in the past are no longer running...? Am I missing anything?
Other than that this looks good.

**Comment by wipfli** [comment]
I haven't seen `build-csp` anywhere. Also I don't know what `prepublishOnly` refers to.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -129,6 +129,37 @@ jobs:
           TAP_COLORS: 1
         run: xvfb-run -s "-ac -screen 0 1280x1024x24" npm run test-browser
 
+  lint_build_test_windows:
```
Comment (HarelM) [inline .github/workflows/ci.yml:132]: This seems like some duplicate of the above.
I think we should either split the jobs by OS or by type (build/lint/test).
But this on is duplicating some of the above without using matrix for OS and some parts already run in other stpes basically (lint_build_test mainly)...

**Comment by HarelM** [comment]
`build-csp` used to create a file in the output folder which might be important for the npm package.
I don't know if and when this was changed... weird... I don't know...
We can also remove the comments at the bottom of the file if we are not planning to uncomment them...

File: .github/workflows/ci.yml (hunk 1 above)
Comment (wipfli) [reply .github/workflows/ci.yml:132]: The only added value here is `test-render`. I wish we could use the os matrix better if only the linux runner had a GPU. Without WebGL we need to run the unit tests on linux in a different way than on macos and windows.

Comment (HarelM) [reply .github/workflows/ci.yml]: I'm not sure I fully follow - we can have either one of the following from my point of view:
1. By OS:
    1. Win - run 1,2,3
    2. Mac - run 1,2,3,4
    3. Linux - run 1,2,4
 2. By step:
    1. lint - win, mac, linux
    2. unit test - win, linux, 
    3. build - win, mac, linux
    4. ...

But currently it's mixed, right?

**Comment by wipfli** [comment]
Why do I have to run `test-unit` before `test-render`?

**Comment by HarelM** [comment]
No, but you have to build before each.

**Comment by wipfli** [comment]
```
> maplibre-gl@2.0.0-pre.3 test-render D:\a\maplibre-gl-js\maplibre-gl-js
> node --experimental-specifier-resolution=node --experimental-json-modules --max-old-space-size=2048 test/render.test.js

(node:2164) ExperimentalWarning: Importing JSON modules is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/D:/a/maplibre-gl-js/maplibre-gl-js/rollup/build/tsc/index' imported from D:\a\maplibre-gl-js\maplibre-gl-js\test\suite_implementation.js
```

**Comment by wipfli** [comment]
Why does this one then pass? https://github.com/maplibre/maplibre-gl-js/pull/335/commits/3f7377ef84d01274d5d2f0ec6b87bbf53c03af68

**Comment by wipfli** [comment]
Something is not coherent here. I suspect it is GitHub not handling stuff right on windows.

**Comment by HarelM** [comment]
You are not running build-tsc before you run the tests in Windows right now...

**Comment by wipfli** [comment]
If https://github.com/maplibre/maplibre-gl-js/pull/335/commits/3f7377ef84d01274d5d2f0ec6b87bbf53c03af68 passes, https://github.com/maplibre/maplibre-gl-js/pull/335/commits/7237d7176d3bfeb146931d26a19528c5141a2b14 should pass too.

**Comment by wipfli** [comment]
> You are not running build-tsc before you run the tests in Windows right now...

We never were.

**Comment by HarelM** [comment]
It might be that other tests did that, you can look into the package.json file to be sure :-)

**Comment by wipfli** [comment]
Does the file system get cleaned before every job or do they share disk space?

**Comment by wipfli** [comment]
> It might be that other tests did that, you can look into the package.json file to be sure :-)

None of them call other scripts.

**Comment by wipfli** [comment]
Beautiful, now `build-tsc` fails: https://github.com/maplibre/maplibre-gl-js/pull/335/checks?check_run_id=3588952281

**Comment by wipfli** [comment]
Errors are not reproducible. See the commits section of this pull request.

**Comment by HarelM** [comment]
I still think you need to build before the tests. Ignore past success commits as I think it's didn't pass successfully, I think it was a glitch...

**Comment by wipfli** [comment]
Ok. Like I did here https://github.com/maplibre/maplibre-gl-js/pull/335/commits/772e970aba1ed8fd43929a54f68476cabfdd9d11 ?

**Comment by HarelM** [comment]
Yes. The errors in the tests seem to indicate that the js files are missing, due to missing build, I think. Try and add back the build and I'll look in the error and see if it changed...

**Comment by wipfli** [comment]
Please see https://github.com/maplibre/maplibre-gl-js/pull/335/checks?check_run_id=3588952281

**Comment by HarelM** [comment]
It's the `now` thing I removed which didn't cause the CI to fall, but this wasn't part of the CI, I guess...
You want me to try and run these tests on a windows machine and fix what's needed?

**Comment by HarelM** [comment]
I have no doubt build is needed, the only question is what else :-)

**Comment by wipfli** [comment]
Harel if you could give it a try on a windows machine it would be great!

**Comment by HarelM** [comment]
This is disappoint I must admit:
I just looked at the release process and it passes although a lot of test fails (Most of them):
https://github.com/maplibre/maplibre-gl-js/runs/3575205638?check_suite_focus=true
`test-render` doesn't pass on windows, it fails with all kind of errors.
I don't know if this was the situation before the migration to typescript, but it is the situation now.
I ran `test-render` on my mac now and it fails with the same error as it fails on windows (`now is not defined`).
So this is not windows specific which is good since we'll be able to fix it in any platform.
Fixing this `now` is easy, but then I get another error about image sizes.
The following are the lines that I removed in `suite_implementation.js`:
```
let now = 0;
browser.now = function() {
    return now;
};
```
I don't understand how now can be used in another method - closure here is confusing, I must be missing out here on something simple...

So `test-render` was never a part of the CI, and didn't really pass when creating a release (This is my conclusion).
I think we should make an effort to fix this, not sure if it is a part of this PR or another one...
Let me know how you'd like to proceed...

**Comment by wipfli** [comment]
Factored out `test-render` into a separate workflow in https://github.com/maplibre/maplibre-gl-js/pull/339.

**Comment by wipfli** [comment]
Removed `test-render` because it will only work with a data provider. The plan is to implement this in a separate pull request.

This update of the release process is ready to be merged in my opinion.

**Review by HarelM — APPROVED** [review]

**Comment by HarelM** [comment]
Worth testing this with a pre.4 release or something...

**Comment by wipfli** [comment]
https://www.npmjs.com/package/maplibre-gl/v/2.0.0-pre.4 looks OK?

**Comment by HarelM** [comment]
Looks good! Good job!

**Pull request merged**